tests/std/tests/P0476R2_bit_cast/test.cpp

```diff
@@ -0,0 +1,271 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifdef __EDG__ // TRANSITION, VSO-1041044
+int main() {}
+#else // __EDG__ ^^^ / vvv !__EDG__
+
+#include <assert.h>
+#include <bit>
+#include <cmath>
+#include <limits>
+#include <stddef.h>
+#include <string.h>
+#include <type_traits>
+#include <utility>
+
+// structure sizing depends on a packing of 8, which is the default
+
+union test_union_1 {
+    char a;
+    int b;
+};
+union test_union_2 {
+    char a;
+    float b;
+};
+
+struct middle_class_1 {
+    double d;
+    virtual void a_member_function_1() {}
+};
+
+struct middle_class_2 {
+    int a;
+    virtual void a_member_function_2() {}
+};
+
+struct derived_class : middle_class_1, middle_class_2 {
+    virtual void a_member_function_2() override {}
+};
+
+struct test_struct_1 {
+    char a;
+    // char[3]
+    int b;
+    short c;
+    // char[6]
+    double d;
+
+    void a_member_function() {}
+};
+
+struct test_struct_2 {
+    short a;
+    // char[2]
+    float b;
+    float c;
+    int d;
+    unsigned int e;
+    char f;
+    // char[1]
+    short g;
+
+    void a_member_function() {}
+};
+
+struct test_struct_3_member_fn_pointer {
+    char a;
+    // char[3]
+    int b;
+    short c;
+    char d[2];
+    void (test_struct_1::*fn)();
+};
+
+struct test_struct_4_large_member_fn_pointer {
+    char a;
+    // char[3]
+    int b;
+    double c;
+    double d;
+    void (derived_class::*fn)(); // "large" member fn pointers are aligned to 8 on both x64 and x86
+};
+
+struct test_struct_5_struct {
+    char a;
+    // char[3]
+    int b;
+    double c;
+    double d;
+    void* e;
+    size_t f;
+};
+
+struct test_struct_6 {
+    char a;
+    // char[3]
+    int b;
+    int c;
+    // char[4] on x64
+    void* v;
+};
+
+struct test_struct_7_member_fn_pointer {
+    char a;
+    // char[3]
+    int b;
+    int c;
+    void (test_struct_1::*fn)();
+};
+
+struct test_struct_1_not_trivially_copyable {
+    char a;
+    // char[3]
+    int b;
+    short c;
+    // char[2]
+    double d;
+
+    test_struct_1_not_trivially_copyable& operator=(const test_struct_1_not_trivially_copyable&) {
+        return *this;
+    }
+};
+
+#pragma pack(push, 1)
+struct test_struct_1_packed {
+    char a;
+    int b;
+    short c;
+    double d;
+};
+struct test_struct_2_packed {
+    short a;
+    float b;
+    float c;
+    int d;
+    unsigned int e;
+    char f;
+    short g;
+};
+#pragma pack(pop)
+
+static_assert(sizeof(test_struct_1) == sizeof(test_struct_2));
+
+template <typename To, typename From, typename = void>
+constexpr bool bit_cast_invocable = false;
+
+template <typename To, typename From>
+constexpr bool bit_cast_invocable<To, From, std::void_t<decltype(std::bit_cast<To>(std::declval<From>()))>> = true;
+
+template <int zero = 0, int = ((void) std::bit_cast<test_union_1>(test_union_2{}), zero)>
+constexpr bool bit_cast_is_constexpr_union(int) {
+    return true;
+}
+
+template <int = 0>

```

Comment:
These do-nothing *template-head*s on the "fallback" `long` overloads seem to be extraneous; we don't pass any explicit template arguments to the overload sets when we use these below. Could we omit these?

Reply:
Omitting them sounds correct. C++ has a rule to prefer non-templated functions over templated functions, but that's a tiebreaker (which is considered in a specific sequence of tiebreakers). Notably, the tiebreaker doesn't apply if a function template is a better match, e.g. by being an exact match for `int` instead of an integral conversion from `int` to `long`.

Reply:
we could, but I don't think it's worth running prtests again.